libvast/vast/system/shutdown.hpp

```diff
@@ -0,0 +1,48 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <caf/actor_cast.hpp>
+#include <caf/event_based_actor.hpp>
+#include <caf/fwd.hpp>
+#include <caf/stateful_actor.hpp>
+
+#include <vector>
+
+namespace vast::policy {
+
+struct sequential;
+struct parallel;
+
+} // namespace vast::policy
+
+namespace vast::system {
+
+/// Performs an asynchronous shutdown of an actor by first shutting down a set
+/// of dependent actors. The shutdown process runs either sequentially or in
+/// parallel. As soon as all dependent actors have terminated, the calling
+/// actor exits.
+/// @param self The actor to terminate.
+/// @param xs Owned actors by *self* that need to shutdown prior to *self*.
+template <class Policy>
+void shutdown(caf::event_based_actor* self, std::vector<caf::actor> xs);
```

Comment:
After thinking about it, I'm not sure this works as you intended: You can call shutdown with any `Policy`, but unless it is `sequential` or `parallel` it should generate a linker error because these are the only ones generated in `shutdown.cpp`.

Probably the correct way would be to say
```
template<class Policy>
void shutdown(...) = delete;

template<>
void shutdown<policy::sequential>(...);


template<>
void shutdown<policy::parallel>(...);
```

(although this is only theorycrafting, I didn't test it with another type)

Reply:
There's a `static_assert` that fires if the policy is neither `sequential` nor `parallel`.

Reply:
I don't see how that can work, the `static_assert` is in a different translation unit so it should not be visible at the call site of, e.g.,  `shutdown<int>()`.

Reply:
This would be the correct way to express the `= delete` here if I recall correctly (not tested):

```cpp
template <class Policy>
void shutdown(...) {
  static_assert(detail::always_false_v<Policy>, "unsupported policy");
}
```